### ISSUE TITLE
Surprising behavior with last in scalaz-stream

### ISSUE THREAD

**Issue by ceedubs** [functional-streams-for-scala/fs2#838]
Note: this pertains to scalaz-stream (0.8 and before). My project hasn't yet made the jump to fs2 (and I haven't yet checked what the behavior there is).

The `last` `process1` seems to emit the last element encountered in a stream even when that stream ends with a failure. This behavior extends to the `fold` `process1` since it uses `scan` and then `last`.

This behavior surprised me (and in fact was the source of a bug in some of my code). Is this intended?

Example:

```scala
scala> val x = Process(1, 2, 3) ++ Process.fail(new RuntimeException("blargh"))
x: scalaz.stream.Process[Nothing,Int] = Append(Emit(WrappedArray(1, 2, 3)),Vector(<function1>))

scala> x.toSource.last.runLog.unsafePerformSyncAttempt // expected output
res0: scalaz.\/[Throwable,Vector[Int]] = -\/(java.lang.RuntimeException: blargh)

scala> x.toSource.last.take(1).runLog.unsafePerformSyncAttempt // shouldn't this be a failure?
res1: scalaz.\/[Throwable,Vector[Int]] = \/-(Vector(3))
```

**Comment by ceedubs** [comment]
@pchlupacek has suggested that this looks similar to #684, which was fixed in 0.8.4, but I'm still seeing this behavior in the latest of the 0.8 series.

**Comment by ceedubs** [comment]
@rossabaker and I just walked through this together and I think that we have a little more insight into what's going on here.
 
Here is the implementation of `last`:

```scala
  def last[I]: Process1[I, I] = {
    def go(prev: I): Process1[I, I] = receive1Or[I,I](emit(prev))(go)
    receive1(go)
  }
```

This emits the previous element when the stream terminates with an `EarlyCause`. This threw me off a bit, because I would expect you to do that when you hit `End` as opposed to an `EarlyCause`, but as I understand it, when piping through `process1`, normal termination is propagated via `Kill`. So the problem ends up being that we handle `Kill` and `Error` the same way with this implementation.

It seems to solve the immediate issue if you change the implementation of `last` to this:

```scala
  def last[I]: Process1[I, I] = {
    def go(prev: I): Process1[I, I] = awaitOr(Get[I]){
      case Cause.Kill => emit(prev)
      case Cause.Error(err) => Process.fail(err)
    }(go)
    receive1(go)
  }
``` 

One thing that I'm not sure about is whether this could result in silent failure in processes that were terminated via `Kill` for something other than normal termination while piping through a `process1`. I definitely don't have a solid understanding of the inner-workings of scalaz-stream -- does anyone who does have thoughts on this? cc @pchlupacek @djspiewak @mpilquist

**Comment by ceedubs** [comment]
For what it's worth, this doesn't seem to be an issue in the 1.0 series.

@mpilquist @pchlupacek if I were to submit a PR with the change that I pasted in my last comment would you have any interest in it? It would change behavior in certain cases -- it seems like people tend to agree that it would change it to more expected behavior, but it's a change in behavior nonetheless.

**Comment by pchlupacek** [comment]
We will not fix any 0.8 issues (#1091). Please migrate to 0.10. We would be happy to help you with migration path. Thank you for understanding.

**Comment by ceedubs** [comment]
@pchlupacek thanks for the followup on this and the offer to help with migration. I am no longer involved in any project using older version of this project, so I should have closed this out earlier.
